components/recorder-cli/default/jest.mjs

```diff
@@ -39,48 +51,67 @@ export const record = (configuration) => {
   }
   const { beforeEach, afterEach, expect } = globalThis;
   const agent = openAgent(configuration);
+  // Jest is claiming that it is running the tests from a given file serially.
+  // However we detected nested calls of beforeEach / afterEach in practice.
+  // So we need to maintain a set of track ids instead of a single one.
+  const tracks = new Map();
+  defineProperty(tracks, "toJSON", {
+    __proto__: null,
+    writable: true,
+    configurable: true,
+    enumerable: true,
+    value: showTrackMap,
+  });
   hook(agent, configuration);
-  let track = null;
-  beforeEach(() => {
+  beforeEach(function () {
     assert(
       !logErrorWhen(
-        track !== null,
-        "Detected concurrent jest test cases %j",
-        track,
+        tracks.has(this),
+        "Duplicate beforeEach test context %o in %j",
+        this,
+        tracks,
       ),
-      "Concurrent jest test cases",
       ExternalAppmapError,
+      "Duplicate beforeEach test context",
     );
     // We cannot use a counter because another jest
     // agent may be running in a different context.
-    track = `jest-${getUuid()}`;
+    const track = `jest-${getUuid()}`;
+    const name = expect.getState().currentTestName;
+    tracks.set(this, { track, name });
+    logWarningWhen(
+      tracks.size > 1,
+      "Detected concurrent jest tests (beforeEach): %j",
+      tracks,
+    );
     recordStartTrack(
       agent,
       track,
-      extendConfiguration(
-        configuration,
-        {
-          "map-name": expect.getState().currentTestName,
-        },
-        null,
-      ),
+      extendConfiguration(configuration, { "map-name": name }, null),
     );
   });
-  afterEach(() => {
+  afterEach(function () {
     assert(
```

Comment:
I see in the commit message you weren't able to reproduce it in a test case.

Does this mean arrow functions within `beforeEach`/`afterEach` won't work since they won't have a unique assignment to `this`?

Reply:
Jest is passing a unique object as `this` to the callbacks of `beforeEach` and `afterEach`. The idea of this PR is to use that object as key to be able to stop the correct track in case of concurrent tests. To catch `this` I need this callback to be a function rather than an arrow.

Reply:
I understand now. I was thinking this also affected end-user `beforeEach` callbacks but I see this is strictly relevant to our own internal setup and teardown.